### ISSUE TITLE
v0.98 syntax-rules doesn't allow ... and _ as matched literals

### ISSUE THREAD

**Issue by WillClinger** [larcenists/larceny#720]
Reported by Alex Shinn:
```
% larceny --r7rs
Larceny v0.98 "General Ripper" (Mar  7 2015 01:06:26, precise:Linux:unified)
larceny.heap, built on Sat Mar  7 01:06:50 EST 2015

> (define-syntax foo
    (syntax-rules (_)
     ((foo _)
      'under)
     ((foo x)
      'other)))

Syntax violation: syntax-case

Invalid form

Form: (syntax-case
  x
  (_)
  ((dummy _) #''under)
  ((dummy x) #''other))

Trace: 

  (syntax-case
  x
  (_)
  ((dummy _) #''under)
  ((dummy x) #''other))

  (lambda (x)
  (syntax-case
    x
    (_)
    ((dummy _) #''under)
    ((dummy x) #''other)))

  (syntax-rules
  (_)
  ((foo _) 'under)
  ((foo x) 'other))

Error: unhandled condition:
Compound condition has these components: 
#<record &who>
    who : syntax-case
#<record &message>
    message : "Invalid form"
#<record &syntax>
    form : (syntax-case x (_) ((dummy _) #''under) ((dummy x) #''other))
    subform : #f

Entering debugger; type "?" for help.
debug> q

> (define-syntax bar
    (syntax-rules (...)
     ((bar ...)
      'dots)
     ((bar x)
      'other)))

Syntax violation: syntax-case

Invalid form

Form: (syntax-case
  x
  (...)
  ((dummy ...) #''dots)
  ((dummy x) #''other))

Trace: 

  (syntax-case
  x
  (...)
  ((dummy ...) #''dots)
  ((dummy x) #''other))

  (lambda (x)
  (syntax-case
    x
    (...)
    ((dummy ...) #''dots)
    ((dummy x) #''other)))

  (syntax-rules
  (...)
  ((bar ...) 'dots)
  ((bar x) 'other))

Error: unhandled condition:
Compound condition has these components: 
#<record &who>
    who : syntax-case
#<record &message>
    message : "Invalid form"
#<record &syntax>
    form : (syntax-case x (...) ((dummy ...) #''dots) ((dummy x) #''other))
    subform : #f

Entering debugger; type "?" for help.
debug> q

> (define-syntax baz
    (syntax-rules (___)
     ((baz ___)
      'under3)
     ((baz x)
      'other)))

> 
```
As Alex remarked, the `foo` and `bar` macros above are invalid in R6 Scheme but valid in R7 Scheme.

This is related to ticket #699.  Adding that extension and fixing this bug will involve a careful rewrite of a central component of Andre van Tonder's macro expander.

**Comment by ashinn** [comment]
Since these work in the R5RS syntax-rules, might it not be simpler to extend that with the R7RS extensions?

**Comment by WillClinger** [comment]
No.  Larceny's implementation of R5RS `syntax-rules` lacks support for several features of R6RS and R7RS `syntax-rules`, and contains several known bugs we don't intend to fix because we anticipate eventual replacement of Larceny's R5RS mode by an extension of Larceny's current `-r7r6` mode.  Even more importantly, Larceny's R5RS `syntax-rules` does not interoperate with the R7RS/R6RS library mechanism.

Andre van Tonder wrote the original library/macro expander that's at the heart of Larceny's R7RS/R6RS modes.  I have a lot of respect for his code, and have tried not to modify it unnecessarily.  The bugs described in tickets #699 and #720 make modifications necessary, but I don't think those mods will be entirely trivial.

**Comment by WillClinger** [comment]
The ellipses issue was fixed when ticket #699 was fixed by changeset 4aec7e64c13edbe67f64a2bcf5ba49b7c9724100

The same approach should fix the underscore bug as well.
